host/frontend/host-orchestrator/operation.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+type NewOperationError string
+
+func (s NewOperationError) Error() string {
+	return fmt.Sprintf("failure creating a new operation: %s", string(s))
+}
+
+type OperationData struct {
+	Name   string
+	Done   bool
+	Result OperationResultData
+}
+
+type OperationResultData struct {
+	Error OperationErrorData
+	OK    OperationOKData
+}
+
+type OperationErrorData struct {
+	ErrorMsg string
+}
+
+type OperationOKData struct{}
+
+type OperationManager interface {
+	NewOperation() (OperationData, error)
+
+	GetOperation(string) (OperationData, bool)
+
+	CompleteOperation(string, OperationResultData)
+}
+
+type MapOM struct {
+	uuidFactory func() string
+	mutex       sync.RWMutex
+	operations  map[string]OperationData
+}
+
+func NewMapOM(uuidFactory func() string) *MapOM {
+	return &MapOM{
+		uuidFactory: uuidFactory,
+		mutex:       sync.RWMutex{},
+		operations:  make(map[string]OperationData),
+	}
+}
+
+func (m *MapOM) NewOperation() (OperationData, error) {
```

Comment:
Creating an operation should never fail, there are infinite valid strings for the operation name and the only reason this fails is because the generated one is already chosen, if that's the case you should generate a new one until an unused one appears.

This would break the test where you use a different factory that always returns the same name, but that's easily fixable: Don't use that factory and get the name from the returned operation, which is guaranteed to have a unique name.

Given that you don't need the factory to ensure a specific name in the test and that you need to guarantee different names for different operations, you should get rid of the factory altogether and use UUID directly from this code.

Reply:
i think being defensive about uiid generation makes the code more robust, why should we trust 100% the uuid generation library, suddenly this library has a bug that returns the same id every time under some special circumstances and we have no protection against it. Having said so, I though of having a retry logic, but still we'll need to test the retry logic so I decided to fail the first time and and have a simpler test as I recognize this could be a very very edge case.